sdks/java/io/solr/src/test/java/org/apache/beam/sdk/io/solr/SolrIOTest.java

```diff
@@ -0,0 +1,263 @@
+import org.junit.rules.ExpectedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A test of {@link SolrIO} on an independent Solr instance.
+ */
+@ThreadLeakScope(value = ThreadLeakScope.Scope.NONE)
+@SolrTestCaseJ4.SuppressSSL
+public class SolrIOTest extends SolrCloudTestCase{
+  private static final Logger LOG = LoggerFactory.getLogger(SolrIOTest.class);
+
+  private static final String SOLR_COLLECTION = "beam";
+  private static final int NUM_SHARDS = 3;
+  private static final long NUM_DOCS = 400L;
+  private static final int NUM_SCIENTISTS = 10;
+  private static final long BATCH_SIZE = 200L;
+
+  private static AuthorizedSolrClient<CloudSolrClient> solrClient;
+  private static SolrIO.ConnectionConfiguration connectionConfiguration;
+
+  @Rule
+  public TestPipeline pipeline = TestPipeline.create();
+
+  @BeforeClass
+  public static void beforeClass() throws Exception {
+    //setup credential for solr user
+    String password = "SolrRocks";
+    byte[] salt = new byte[32];
+    String base64Salt = BaseEncoding.base64().encode(salt);
+    String sha56 = Sha256AuthenticationProvider.sha256(password, base64Salt);
+    String credential = sha56 + " " + base64Salt;
+    String securityJson = "{"
+        + "'authentication':{"
+        + "  'blockUnknown': true,"
+        + "  'class':'solr.BasicAuthPlugin',"
+        + "  'credentials':{'solr':'" + credential + "'}}"
+        + "}";
+
+    configureCluster(3)
+        .addConfig("conf", getFile("cloud-minimal/conf").toPath())
+        .configure();
+    ZkStateReader zkStateReader = cluster.getSolrClient().getZkStateReader();
+    zkStateReader.getZkClient()
+        .setData("/security.json", securityJson.getBytes(Charset.defaultCharset()), true);
+    String zkAddress = cluster.getZkServer().getZkAddress();
+    connectionConfiguration = SolrIO.ConnectionConfiguration.create(zkAddress, SOLR_COLLECTION)
+        .withBasicCredentials("solr", password);
+    solrClient = connectionConfiguration.createClient();
+    SolrIOTestUtils.createCollection(SOLR_COLLECTION, NUM_SHARDS, 1, solrClient);
+  }
+
+  @AfterClass
+  public static void afterClass() throws Exception {
+    solrClient.close();
+  }
+
+  @Before
+  public void before() throws Exception {
+    SolrIOTestUtils.clearCollection(solrClient);
+  }
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  public void testBadCredentials() throws IOException {
+    thrown.expect(SolrException.class);
+
+    String zkAddress = cluster.getZkServer().getZkAddress();
+    SolrIO.ConnectionConfiguration connectionConfiguration = SolrIO.ConnectionConfiguration
+        .create(zkAddress, SOLR_COLLECTION)
+        .withBasicCredentials("solr", "wrongpassword");
+    try (AuthorizedSolrClient solrClient = connectionConfiguration.createClient()) {
+      SolrIOTestUtils.insertTestDocuments(NUM_DOCS, solrClient);
+    }
+  }
+
+  @Test
+  public void testSizes() throws Exception {
```

Comment:
Looking at https://beam.apache.org/documentation/io/testing/, I think SolrIOIT should be quite different. I didn't realize how much guidance we had on writing integration tests! Sorry for not consulting that document earlier. Please take a look at that document, section "I/O Transform Integration Tests" and in particular "Implementing Integration Tests"? I think in practice it means, convert SolrIOIT to be a single "write then read" test, using `TestRow`, using unique collection names etc. I think also it may be a good idea to set that aside for a separate PR, because the rest of the current PR is looking good and I'm feeling fairly safe about shipping SolrIO with just the unit tests, and adding integration test separately, because the unit tests use a real in-memory instance and give pretty good coverage.